ddmark/ddmark.go

```diff
@@ -0,0 +1,185 @@
+	"reflect"
+
+	k8sloader "sigs.k8s.io/controller-tools/pkg/loader"
+	k8smarkers "sigs.k8s.io/controller-tools/pkg/markers"
+)
+
+func InitializeMarkers() *k8smarkers.Collector {
+	col := &k8smarkers.Collector{}
+	reg := &k8smarkers.Registry{}
+
+	// takes all the markers definition found in the ddmark/validation package, prior to analyzing the packages
+	err := Register(reg)
+	if err != nil {
+		fmt.Printf("\nerror loading markers from crd validation: %v", err)
+		return col
+	}
+
+	col.Registry = reg
+
+	return col
+}
+
+// ValidateStruct applies struct markers found in structPkgs struct definitions to a marshalledStruct object.
+// It allows to enforce markers rule onto that object, according to the constraints defined in structPkgs
+func ValidateStruct(marshalledStruct interface{}, filePath string, structPkgs ...string) []error {
+	col := InitializeMarkers()
+
+	var err error
+
+	var errorList []error = make([]error, 0)
+
+	var pkgs []*k8sloader.Package
+
+	pkgs, err = k8sloader.LoadRoots(structPkgs...)
+
+	if err != nil {
+		return append(errorList, fmt.Errorf("error loading markers from crd validation: \n\t%v", err))
+	}
+
+	typesMap := getAllPackageTypes(pkgs, col)
+	if len(typesMap) == 0 {
+		errorList = append(errorList, fmt.Errorf("%v: loaded classes are empty or not found", filePath))
+	}
+
+	validateStruct(marshalledStruct, typesMap, nil, &errorList, filePath, col)
+
+	return errorList
+}
+
+// validateStruct is an internal recursive function that recursively applies markers rules to types and fields
+func validateStruct(marshalledStruct interface{}, typesMap map[string]*k8smarkers.TypeInfo, markerValues k8smarkers.MarkerValues, errorList *[]error, fieldName string, col *k8smarkers.Collector) {
+	value := reflect.Indirect(reflect.ValueOf(marshalledStruct)) // dereferences pointer value if there is one
+	if value.IsValid() && !value.IsZero() {
+		markerType := typesMap[value.Type().Name()]
+		if markerType != nil {
+			applyMarkers(value, markerType.Markers, errorList, fieldName, k8smarkers.DescribesType, col)
+
+			// apply markers to each fields
+			for _, field := range markerType.Fields {
+				if fieldValue := value.FieldByName(field.Name); fieldValue.IsValid() {
+					validateStruct(
+						fieldValue.Interface(),
+						typesMap,
+						field.Markers,
+						errorList,
+						fieldName+">"+field.Name,
+						col,
+					)
+				}
+			}
+		}
+
+		if value.Kind() == reflect.Slice || value.Kind() == reflect.Array {
+			for i := 0; i < value.Len(); i++ {
+				validateStruct(value.Index(i).Interface(), typesMap, nil, errorList, fieldName+">"+value.Type().Name(), col)
+			}
+		}
+	}
+
+	applyMarkers(value, markerValues, errorList, fieldName, k8smarkers.DescribesField, col)
+}
+
+// applyMarkers applies all markers found in the markers arg to a given type/field
+func applyMarkers(value reflect.Value, markers k8smarkers.MarkerValues, errorList *[]error, fieldName string, targetType k8smarkers.TargetType, col *k8smarkers.Collector) {
```

Comment:
This function is logically laid out, but could you comment the inside of it a bit more, just to make it smoother to read?

Reply:
Comments added, lmk if something still isn't clear to you !